js/modules/k6/grpc/client.go

```diff
@@ -0,0 +1,479 @@
+	addr string, in net.Conn) (net.Conn, credentials.AuthInfo, error) {
+	out, auth, err := t.TransportCredentials.ClientHandshake(ctx, addr, in)
+	if err != nil {
+		t.errc <- err
+	}
+
+	return out, auth, err
+}
+
+// Connect is a block dial to the gRPC server at the given address (host:port)
+func (c *Client) Connect(ctxPtr *context.Context, addr string, params map[string]interface{}) (bool, error) {
+	state := lib.GetState(*ctxPtr)
+	if state == nil {
+		return false, errConnectInInitContext
+	}
+
+	isPlaintext, timeout := false, 60*time.Second
+
+	for k, v := range params {
+		switch k {
+		case "plaintext":
+			isPlaintext, _ = v.(bool)
+		case "timeout":
+			if t, ok := v.(float64); ok && t > 0.0 {
+				timeout = time.Duration(t) * time.Millisecond
+			}
+		}
+	}
+
+	errc := make(chan error)
+	go func() {
+		opts := []grpc.DialOption{
+			grpc.WithBlock(),
+			grpc.FailOnNonTempDialError(true),
+			grpc.WithStatsHandler(c),
+		}
+
+		if ua := state.Options.UserAgent; ua.Valid {
+			opts = append(opts, grpc.WithUserAgent(ua.ValueOrZero()))
+		}
+
+		if !isPlaintext {
+			tlsCfg := state.TLSConfig.Clone()
+			tlsCfg.NextProtos = []string{"h2"}
+
+			// TODO(rogchap): Would be good to add support for custom RootCAs (self signed)
+
+			// (rogchap) we create a wrapper for transport credentials so that we can report
+			// on any TLS errors.
+			creds := transportCreds{
+				credentials.NewTLS(tlsCfg),
+				errc,
+			}
+			opts = append(opts, grpc.WithTransportCredentials(creds))
+		}
+
+		if isPlaintext {
+			opts = append(opts, grpc.WithInsecure())
+		}
+
+		dialer := func(ctx context.Context, addr string) (net.Conn, error) {
+			return state.Dialer.DialContext(ctx, "tcp", addr)
+		}
+		opts = append(opts, grpc.WithContextDialer(dialer))
+
+		ctx, cancel := context.WithTimeout(*ctxPtr, timeout)
+		defer cancel()
+
+		var err error
+		c.conn, err = grpc.DialContext(ctx, addr, opts...)
+		if err != nil {
+			errc <- err
+
+			return
+		}
+		close(errc)
+	}()
+
+	if err := <-errc; err != nil {
+		return false, err
+	}
+
+	return true, nil
+}
+
+// InvokeRPC creates and calls a unary RPC by fully qualified method name
+//nolint: funlen,gocognit
+func (c *Client) InvokeRPC(ctxPtr *context.Context,
+	method string, req goja.Value, params map[string]interface{}) (*Response, error) {
+	ctx := *ctxPtr
+	rt := common.GetRuntime(ctx)
+	state := lib.GetState(ctx)
+	if state == nil {
+		return nil, errInvokeRPCInInitContext
+	}
+
+	if c.conn == nil {
+		return nil, errors.New("no gRPC connection, you must call connect first")
+	}
+
+	method = strings.TrimPrefix(method, "/")
+	md := c.mds[method]
+
+	if md == nil {
+		return nil, fmt.Errorf("method %q not found in file descriptors", method)
+	}
+
+	c.tags = state.CloneTags()
+	timeout := 60 * time.Second
+
+	ctx = metadata.NewOutgoingContext(ctx, metadata.New(nil))
+	for k, v := range params {
+		switch k {
+		case "headers":
+			rawHeaders, ok := v.(map[string]interface{})
+			if !ok {
+				continue
```

Comment:
A `struct` would make unmarshaling the `timeout` parameter challenging now that we support both number and string duration.

Reply:
Not necessarily, see how our own `Duration` wrapper handles it both integer and string values: https://github.com/loadimpact/k6/blob/4e637f6e57906584a5ae2f90da836802be4e03c9/lib/types/types.go#L137

That said, it still doesn't handle "int duration is in ms", since https://github.com/loadimpact/k6/issues/1305 is a minor breaking change for a lot of things. Hopefully we'll fix that in k6 v0.30.0, but we'll see. For now this is perfectly fine, we can refactor it in the future, if we need or want to.